geopmdpy/geopmdpy/session.py

```diff
@@ -300,16 +308,33 @@ def run(self, run_time, period, pid, print_header,
             header_names = self.header_names(signal_config)
             print(self._delimiter.join(header_names), file=out_stream)
         with stats.Collector(signal_config) if do_stats else nullcontext() as stats_collector:
-            self.run_read(requests, run_time, period, pid, out_stream, stats_collector)
-            if do_stats:
-                report = stats_collector.report_yaml()
-        if do_stats:
-            if session_io is not None:
-                session_io.write_report(report)
+            try:
+                g_session_handler = SessionHandler(do_stats, stats_collector, out_stream, report_path, session_io)
+                self.run_read(requests, run_time, period, pid, out_stream, stats_collector)
+            finally:
+                g_session_handler.stop()
+                g_session_handler = None
+
+class SessionHandler:
+    def __init__(self, do_stats, stats_collector, out_stream, report_path, session_io):
+        self.do_stats = do_stats
+        self.stats_collector = stats_collector
+        self.out_stream = out_stream
+        self.report_path = report_path
+        self.session_io = session_io
+
+    def stop(self):
+        if self.out_stream is not None:
+            self.out_stream.flush()
+        if self.do_stats:
+            report = self.stats_collector.report_yaml()
+            if self.session_io is not None:
+                self.session_io.write_report(report)
```

Comment:
I don't see ``close_trace_stream()`` being called anywhere.  Should it be here?  This was the only spot I found where there's no ``with open()`` context manager, so might be a file left open.

Reply:
close_trace_stream() was dead code, and actually not correctly implemented.  I have put the method in the run path and fixed it.